### ISSUE TITLE
Improved initial Outlines experience

### ISSUE THREAD

**Issue by cpfiffer** [dottxt-ai/outlines#1149]
There's a few areas where the initial experience of using Outlines is a little clunky. 

## Installing Outlines

For one, none of the inference backends [are dependencies](https://github.com/dottxt-ai/outlines/blob/0b9a3f1378c62449472cd6430d18eab9ac39c2bd/pyproject.toml#L26-L44), which means people are usually confronted by a series of "please install X", "please install Y", etc. 

Eric Ma has a [Gist about this](https://gist.github.com/ericmjl/085d9efb7b6ab46383ab19fbd09805ef), which highlights a series of mild irritations as the user's very first interactions with Outlines. Ideally, there should be a set of install options that provide canned or default options for users.

My proposal is to add `llama_cpp` as a default install, and provide a series of optional dependencies by inference engine

```bash
# prepackaged with llamacpp
pip install outlines 

# One for each engine
pip install 'outlines[vllm]'
pip install 'outlines[transformers]'
pip install 'outlines[all]' # gives you everything
```

EDIT: we [already support](https://github.com/dottxt-ai/outlines/blob/0b9a3f1378c62449472cd6430d18eab9ac39c2bd/pyproject.toml#L68-L73) the `serve` extra, so it should be relatively straightforward to include other methods.

## Docs

The docs should also be updated to include examples for a handful of different backends in tabs, so that people can enter their preferred inference mode and have all the docs reflect their choice. 

It may also be able to allow users to select a model and preferred quantization method so they can always copy/paste the code in whatever mode they're working in.

## Models to use and quantization

There's lots of annoying little edge cases with models. Some of the docs still refer to gated models. These are generally pretty reliable, but it requires you to set an HF token which many users may not have already done. 

We should make it very clear what models are recommended as intro tools, and show clearly how to quantize them for resource-constrained users.

Any other things I missed?

**Comment by rlouf** [comment]
I agree on all points. You can split this in smaller issues, and people can keep sharing their grievances here :)

**Comment by lapp0** [comment]
Good issue, I've run into all of these problems.

I disagree about `llama.cpp` though, there's no reason to by default include `llama-cpp-python` in downstream dependents such as vLLM. Additionally, `llama-cpp-python` is the Outlines-supported model I've run into the most issues installing / building.

We should prominently document that users can install `outlines[llamacpp]`, `outlines[transformers]`, etc, rather than recommending `pip install outlines`

**Comment by cpfiffer** [comment]
> I disagree about `llama.cpp` though, there's no reason to by default include `llama-cpp-python` in downstream dependents such as vLLM.

> We should prominently document that users can install `outlines[llamacpp]`, `outlines[transformers]`, etc, rather than recommending `pip install outlines`

This is fine with me!



> Additionally, `llama-cpp-python` is the Outlines-supported model I've run into the most issues installing / building.

I've had this happen as well. Lots of tiny issues with llamacpp. This is a bigger issue IMO -- llamacpp support should definitely be more robust to support the various hardware-constrained Outlines users.

**Comment by rlouf** [comment]
As long as major inference libraries import `outlines` we shouldn't include any library by default. _However_ when the structured generation logic is in a separate library I am ready to consider having `llama.cpp` as a default.

**Comment by ahmed-moubtahij** [comment]
Coming from pure `pip install outlines` (it didn't prompt me to install anything else) it took 1hr+ to generate 512 tokens constrained to a `r"```latex(.*?|\n)```"` regex. The FSM compiled to 100% fairly fast though. This was a 2B-4bit model on a 3090, of which all 24GB of VRAM were filled during generation (my prompt is like 20 tokens).

I had a similar experience in the past, which I "solved" by using HuggingFace's TGI for structured generation. It was a lot faster, which is weird because I thought they used Outlines under the hood.

Should I have went with an inference engine like `pip install outlines[vllm]`?

**Comment by lapp0** [comment]
@ahmed-moubtahij Yes, outlines only becomes the bottleneck after ~1,000 tokens/s, and `vllm` is substantially faster than `transformers`

However, are you sure you set `device_map="cuda"`? Sounds like you might have been generating on CPU.

**Comment by ahmed-moubtahij** [comment]
@lapp0 Yeah I already had device_map set to cuda. Its entire VRAM was occupied.

I'll re-try with `pip install outlines[vllm]`

**Comment by rlouf** [comment]
Do you mind moving the resolution of this problem to a new issue? 🙂

**Comment by ahmed-moubtahij** [comment]
> Do you mind moving the resolution of this problem to a new issue? 🙂

Sure, how do I accomplish that? I thought of referencing my messages in a new issue and deleting from here but I think that would cancel the reference.

**Comment by rlouf** [comment]
You don't have to delete the messages, we'll just hide them here!

**Comment by ahmed-moubtahij** [comment]
Done: https://github.com/dottxt-ai/outlines/issues/1167
